rpc: add hidden getrawaddrman RPC to list addrman table entries

0xB10C opened the pull request:
Inspired by `getaddrmaninfo` (#27511), this adds a hidden/test-only `getrawaddrman` RPC. The RPC returns information on all addresses in the address manager new and tried tables. Addrman table contents can be used in tests and during development. 

The RPC result encodes the `bucket` and `position`, the internal location of addresses in the tables, in the address object's string key. This allows users to choose to consume or to ignore the location information. If the internals of the address manager implementation change, the location encoding might change too.  

```
getrawaddrman

EXPERIMENTAL warning: this call may be changed in future releases.

Returns information on all address manager entries for the new and tried tables.

Result:
{                                  (json object)
  "table" : {                      (json object) buckets with addresses in the address manager table ( new, tried )
    "bucket/position" : {          (json object) the location in the address manager table (<bucket>/<position>)
      "address" : "str",           (string) The address of the node
      "port" : n,                  (numeric) The port number of the node
      "network" : "str",           (string) The network (ipv4, ipv6, onion, i2p, cjdns) of the address
      "services" : n,              (numeric) The services offered by the node
      "time" : xxx,                (numeric) The UNIX epoch time when the node was last seen
      "source" : "str",            (string) The address that relayed the address to us
      "source_network" : "str"     (string) The network (ipv4, ipv6, onion, i2p, cjdns) of the source address
    },
    ...
  },
  ...
}

Examples:
> bitcoin-cli getrawaddrman
> curl --user myusername --data-binary '{"jsonrpc": "1.0", "id": "curltest", "method": "getrawaddrman", "params": []}' -H 'content-type: text/plain;' http://127.0.0.1:8332/
```

0xB10C committed:
rpc: getrawaddrman for addrman entries

Exposing address manager table entries in a hidden RPC allows to introspect
addrman tables in tests and during development.

As response JSON object the following FORMAT1 is choosen:
{
  "table": {
    "<bucket>/<position>": { "address": "..", "port": .., ... },
    "<bucket>/<position>": { "address": "..", "port": .., ... },
    "<bucket>/<position>": { "address": "..", "port": .., ... },
    ...
  }
}

An alternative would be FORMAT2
{
  "table": {
    "bucket": {
      "position": { "address": "..", "port": .., ... },
      "position": { "address": "..", "port": .., ... },
      ..
    },
    "bucket": {
      "position": { "address": "..", "port": .., ... },
      ..
    },
  }
}

FORMAT1 and FORMAT2 have different encodings for the location of the
address in the address manager. While FORMAT2 might be easier to process
for downstream tools, it also mimics internal addrman mappings, which
might change at some point. Users not interested in the address location
can ignore the location key. They don't have to adapt to a new RPC
response format, when the internal addrman layout changes. Additionally,
FORMAT1 is also slightly easier to to iterate in downstream tools. The
RPC response-building implemenation complexcity is lower with FORMAT1
as we can more easily build a "<bucket>/<position>" key than a multiple
"bucket" objects with multiple "position" objects (FORMAT2).

0xB10C committed:
test: getrawaddrman RPC

Test that the getrawaddrman returns the addresses in the new and tried
tables. We can't check the buckets and positions as these are not
deterministic (yet).

brunoerg commented:
Concept ACK

brunoerg commented on src/addrman.cpp:
```diff
@@ -838,6 +838,25 @@ std::vector<CAddress> AddrManImpl::GetAddr_(size_t max_addresses, size_t max_pct
+{
+    AssertLockHeld(cs);
+
+    const int bucket_count = (from_tried) ? ADDRMAN_TRIED_BUCKET_COUNT : ADDRMAN_NEW_BUCKET_COUNT;
```
nit:
```suggestion
    const int bucket_count = from_tried ? ADDRMAN_TRIED_BUCKET_COUNT : ADDRMAN_NEW_BUCKET_COUNT;
```

brunoerg commented on src/rpc/net.cpp:
```diff
@@ -1014,52 +1015,105 @@ static RPCHelpMan sendmsgtopeer()
+                {
+                    {RPCResult::Type::ELISION, "", "The same output as verbose = false"},
+                    {
+                        RPCResult::Type::ARR, "table", "list of addresses in a address manager table ( new_table, tried_table )",
```
```suggestion
                        RPCResult::Type::ARR, "table", "list of addresses in the address manager table ( new_table, tried_table )",
```

0xB10C commented:
- addressed @brunoerg's review
- included  https://github.com/bitcoin/bitcoin/pull/27511#discussion_r1335084087 as I'm touching the line anyway


I've also been hacking on a tool that visualizes addrman tables from the getaddrmaninfo verbose output in the browser. A wip version is on https://0xb10c.github.io/addrman-observer ([github.com/0xB10C/addrman-observer](https://github.com/0xB10C/addrman-observer)). You can load dumps produced with `bitcoin-cli getaddrmaninfo true > getaddrmaninfo.json` (all processing happens locally).

brunoerg commented on src/rpc/net.cpp:
```diff
@@ -1014,52 +1015,105 @@ static RPCHelpMan sendmsgtopeer()
+        RPCExamples{
+            HelpExampleCli("getaddrmaninfo", "")
+            + HelpExampleCli("getaddrmaninfo", "true")
+            + HelpExampleRpc("getaddrmaninfo", "")},
```
In 56adcf42805b3007b66396754a9e5dcd0968d027: We could add the 'verbose' example for the RPC.
```suggestion
            + HelpExampleRpc("getaddrmaninfo", "")},
            + HelpExampleRpc("getaddrmaninfo", "true")}
```

0xB10C replied:
done!

brunoerg reviewed (approved):
light ACK da44fd3e499ca53242332ef9cb2907ec31fce1e3

I intend to do more tests soon. Just tested it on a recent clearnet node and worked as expected.
Outputs: https://gist.github.com/brunoerg/e3643337b4c16923ee0e0f7e66bec4d9

brunoerg commented on test/functional/rpc_net.py:
```diff
@@ -386,5 +391,70 @@ def test_getaddrmaninfo(self):
+        assert 0 <= res["tried_table"][0]["bucket"] < ADDRMAN_TRIED_BUCKET_COUNT
+        assert 0 <= res["tried_table"][0]["position"] < ADDRMAN_BUCKET_SIZE
+
+        self.log.debug("Add one new address to each addrman table")
```
In da44fd3e499ca53242332ef9cb2907ec31fce1e3: I think we could addresses from other networks (e.g tor) to test it as well.

0xB10C replied:
will do if I re-touch the tests

ajtowns commented on src/rpc/net.cpp:
```diff
@@ -1014,52 +1015,107 @@ static RPCHelpMan sendmsgtopeer()
+                        RPCResult::Type::ARR, "table", "list of addresses in the address manager table ( new_table, tried_table )",
+                        {
+                            {RPCResult::Type::OBJ, "", "an address manager table entry", {
+                                {RPCResult::Type::STR, "address", "the address"},
```
`getnodeaddresses` separates this into `address` and `port`; does it make sense to be different here?

`getnodeaddresses` also includes the last seen time which seems like it might be useful?

ajtowns commented on src/rpc/net.cpp:
```diff
@@ -1014,52 +1015,107 @@ static RPCHelpMan sendmsgtopeer()
+                                {RPCResult::Type::STR, "address", "the address"},
+                                {RPCResult::Type::NUM, "services", "the services the node might support"},
+                                {RPCResult::Type::NUM, "bucket", "the address manager bucket the address is placed in"},
+                                {RPCResult::Type::NUM, "position", "the bucket position the address is placed in"},
```
Could substantially reduce the size of the json output (by about 20%) by changing this to:

```json
{
  "new_table": {
    "0/2": {
      "address": "89.78.111.197:8333",
      "services": 1037,
      "source": "34.126.67.135"
    },
    "0/5": {
      "address": "13.228.211.83:8333",
      "services": 1033,
      "source": "34.126.67.135"
    }
}
```

0xB10C replied:
TIL UniValue doesn't support integer keys - only string keys. Still going with this as it's about a 15% size reduction with somewhat empty addrman tables. I assume even more space savings on a full addrman tables, as the output would contain more `bucket` keys.

```
$ ll getrawaddrman.json getaddrmaninfo_verbose.json -h
-rw-r--r-- 1 user group 761K Sep 28 13:37 getrawaddrman.json
-rw-r--r-- 1 user group 881K Sep 28 13:37 getaddrmaninfo_verbose.json
```

ajtowns replied:
JSON in general  doesn't support integer keys.

Not a strong opinion, but there's two advantages of `"0/5":` vs `"0": {"5":` -- one is that if you don't care about the bucket/position, you can just skip through the values directly (`x.values()` vs `sum(x2.values() for x2 in x.values())`), the other is that bucket+position are pretty internal arrangements; if we end up with a different layout in the future, we can just map that to a different string template (eg "0/5/2") and only the people who care about the internals have to even notice the change.

ajtowns commented on src/rpc/net.cpp:
```diff
@@ -1014,52 +1015,107 @@ static RPCHelpMan sendmsgtopeer()
+            bool verbose = false;
+            if (!request.params[0].isNull()) {
+                verbose = request.params[0].get_bool();
+            }
```
Add

```c++
TMPL_INST(CheckRequiredOrDefault, bool, CHECK_NONFATAL(maybe_arg)->get_bool(););
```

in rpc/util.cpp and this becomes `const bool verbose = self.Arg<bool>(0);`  (see #28230)

0xB10C replied:
I choose to move to `getrawaddrman` instead of extending `getaddrmaninfo` with a verbose flag. https://github.com/bitcoin/bitcoin/pull/28523#discussion_r1339454178. Resolving this as there's no need for it anymore.

ajtowns commented on src/rpc/net.cpp:
```diff
@@ -1014,52 +1015,107 @@ static RPCHelpMan sendmsgtopeer()
+            obj.pushKV("total", node.addrman->Size());
+            ret.pushKV("all_networks", obj);
+
+            if (verbose) {
```
We have `getmempoolinfo` that gives statistics about the mempool and `getrawmempool` that gives the contents of the mempool; wouldn't it make more sense to mirror that arrangement and have two separate RPCs here as well?

ajtowns reviewed (commented):
Concept ACK

0xB10C commented:
Thanks for the helpful review comments @brunoerg @ajtowns @amitiuttarwar. I think I have addressed them all in the most recent push:

- https://github.com/bitcoin/bitcoin/pull/28523#discussion_r1337326653: I'm now adding an `onion` and an `ipv6` address (instead of two `ipv4`) to the addrman tables. There are already two IPv4 addresses in the addrman during the test.
- https://github.com/bitcoin/bitcoin/pull/28523#discussion_r1339430595: Inspired by `getnodeaddresses` I added a `time`, `port`, `network`, and `source_network` field. This increases the response size. However, this is highly compressible if someone wants to store or serve it.
- https://github.com/bitcoin/bitcoin/pull/28523#discussion_r1339444441: adopted the `<bucket>/<position>` key. I've implemented both discussed approaches and found `<bucket>/<position>` to be better. See commit message for details.
- https://github.com/bitcoin/bitcoin/pull/28523#discussion_r1339454178: moved this out of `getaddrmaninfo verbose` and into a separate, hidden `getrawaddrman` RPC. This also means https://github.com/bitcoin/bitcoin/pull/27511#discussion_r1335084087 is not included anymore.
- https://github.com/bitcoin/bitcoin/pull/28523#discussion_r1339908447: using `AddressPosition` in a `std::pair<AddrInfo, AddressPosition>` now. This should be easier to reason about than a `std::tuple<int, int, AddrInfo>`.

I've also updated OP and the tool on https://0xb10c.github.io/addrman-observer.

stratospher reviewed (commented):
tested ACK 45a84d1. Very cool to have!

amitiuttarwar reviewed (commented):
tested ACK 45a84d19dec. All the review comments are small style/doc things. The functionality looks good.

ajtowns commented on src/rpc/net.cpp:
```diff
@@ -1063,6 +1064,72 @@ static RPCHelpMan getaddrmaninfo()
+    return RPCHelpMan{
+        "getrawaddrman",
+        "\nReturns information on all address manager entries for the new and tried tables.\n"
+        "This RPC is for testing only.\n",
```
Perhaps it would be better to describe it as "EXPERIMENTAL warning: this call may be changed in future releases." like `sendall` ?

0xB10C commented:
Addresses the comments and nits. Thanks @amitiuttarwar, @stratospher, and @ajtowns.

stratospher commented:
reACK 584e12d.

0xB10C commented on test/functional/rpc_net.py:
```diff
@@ -26,6 +26,11 @@
 )
 from test_framework.wallet import MiniWallet
 
+# Address manager size constants as defined in addrman_impl.h
+ADDRMAN_NEW_BUCKET_COUNT = 1 << 10
+ADDRMAN_TRIED_BUCKET_COUNT = 1 << 8
+ADDRMAN_BUCKET_SIZE = 1 << 6
+
```
If I re-touch this, I'll move the constants to `test/functional/test_framework/netutil.py` as in https://github.com/bitcoin/bitcoin/pull/28176#issuecomment-1741842516

ismaelsadeeq replied:
#28176 was merged, you might want to rebase on it to avoid duplication.

0xB10C commented:
The red CI job with the failing `wallet_upgradewallet.py --legacy-wallet` test looks unrelated.

willcl-ark commented:
This seems awfully slow to execute on my machine, taking about 30 seconds to return...

It appears after some (light) investigation that the time is spent inside `AddrmanTableToJSON`: https://tmp.256k1.dev/23_10_bitcoin_getrawaddrman_split.svg

The working call to (addrman) `GetEntries` is this sliver here, with _all_ the rest of the time taken by processing the result to JSON:

![image](https://github.com/bitcoin/bitcoin/assets/6606587/c835fbe1-7f5f-4f51-bfaf-37f3f8119e7c)

If this is the best we can do then it might not be a blocker for me, as at least we aren't holding locks for all this time, but it does seem a shame to spend 30 seconds processing a near-instant internal call, especially for an RPC which could conceivably be used in a polling situation...

Of course it's also possible that this is just something wrong with my system?

Addrman numbers for reference:

```
$ time /home/will/src/bitcoin/src/bitcoin-cli getaddrmaninfo
{
  "ipv4": {
    "new": 40403,
    "tried": 1623,
    "total": 42026
  },
  "ipv6": {
    "new": 9806,
    "tried": 0,
    "total": 9806
  },
  "onion": {
    "new": 11200,
    "tried": 818,
    "total": 12018
  },
  "i2p": {
    "new": 746,
    "tried": 107,
    "total": 853
  },
  "cjdns": {
    "new": 0,
    "tried": 0,
    "total": 0
  },
  "all_networks": {
    "new": 62155,
    "tried": 2548,
    "total": 64703
  }
}

________________________________________________________
Executed in    8.06 millis    fish           external
   usr time    4.12 millis  723.00 micros    3.39 millis
   sys time    3.58 millis  188.00 micros    3.39 millis

```

0xB10C commented:
Thanks @willcl-ark for testing. I'm seeing about 6s for `time bitcoin-cli getrawaddrman > /dev/null` with similar sized addrman tables ( "new": 61469, "tried": 1157 vs  "new": 62155, "tried": 2548,) on a medium-spec machine. I wonder where the 5x increase in processing time on your side comes from. Can you reproduce this on a different machine?

However, using `perf record`, my flamegraph shows that were spending quite a bit of time in `UniValue::findKey()`. This is O(n) for the number of addrman table entries (here mostly new table entries). The same problem was fixed for `getrawmempool` in #14984 by using `__pushKV()` (now called `pushKVEnd()`), which is O(1). We can use it here too, as our keys are unique.

This reduces `time bitcoin-cli getrawaddrman > /dev/null` from 6s to 900ms for me. I've pushed this as I think this is generally a nice improvement. Would be curious to hear about the improvement on your side.

willcl-ark commented:
> Would be curious to hear about the improvement on your side.

Thanks for the quick response. This certainly seems to give a good-enough improvement on my side:

```
will@ubuntu in ~/src/bitcoin on  2023-09-verbose-getaddrmaninfo [$?] : C v16.0.6-clang : 🐍 3.8.17
₿ git log --oneline | head -n1; and time ./src/bitcoin-cli getrawaddrman > /dev/null
584e12d977 test: getrawaddrman RPC

________________________________________________________
Executed in   30.71 secs    fish           external
   usr time    1.04 secs  204.00 micros    1.04 secs
   sys time    0.07 secs   41.00 micros    0.07 secs


will@ubuntu in ~/src/bitcoin on  2023-09-verbose-getaddrmaninfo [$?] : C v16.0.6-clang : 🐍 3.8.17 took 30s
₿ git log --oneline | head -n1; and time ./src/bitcoin-cli getrawaddrman > /dev/null
2dc1a6d59e test: getrawaddrman RPC

________________________________________________________
Executed in    3.19 secs    fish           external
   usr time    1.09 secs  419.00 micros    1.09 secs
   sys time    0.04 secs   89.00 micros    0.04 secs

```

willcl-ark commented:
tACK 2dc1a6d59e

Nice introspection tool which is very useful for monitoring addrman buckets without the need for tracepoints.

brunoerg reviewed (approved):
crACK 2dc1a6d59e9fed4297d10ae2b4b0b00084538348

0xB10C commented:
With https://github.com/bitcoin/bitcoin/pull/28176 merged, I've rebased and now use the addrman table size constants  in `netutil.py` in the test framework. This also resolves https://github.com/bitcoin/bitcoin/pull/28523#discussion_r1342109520. Sorry for invalidating the ACKs again.

willcl-ark commented:
reACK 352d5eb2a9

git range-diff shows the only difference since previous ACK @ 2dc1a6d is using the just-merged constants from `netutil.py`:


<details>
<summary>Details</summary>

```
₿ git range-diff 2dc1a6d...352d5eb
 -:  ---------- >  1:  63e90e1d3f test: check for specific disconnect reasons in p2p_blockfilters.py
 -:  ---------- >  2:  2ab7952bda test: add bip157 coverage for (start height > stop height) disconnect
 -:  ---------- >  3:  5671c15f45 blockstorage: Mark FindBlockPos as nodiscard
 -:  ---------- >  4:  f0207e0030 blockstorage: Return on fatal block file flush error
 -:  ---------- >  5:  d8041d4e04 blockstorage: Return on fatal undo file flush error
 -:  ---------- >  6:  925bb723ca [refactor] batch-add transactions to DisconnectedBlockTransactions
 -:  ---------- >  7:  59a35a7398 [bench] DisconnectedBlockTransactions
 -:  ---------- >  8:  79ce9f0aa4 add std::list to memusage
 -:  ---------- >  9:  2765d6f343 rewrite DisconnectedBlockTransactions as a list + map
 -:  ---------- > 10:  cf5f1faa03 MOVEONLY: DisconnectedBlockTransactions to its own file
 -:  ---------- > 11:  4313c77400 make DisconnectedBlockTransactions responsible for its own memory management
 -:  ---------- > 12:  fa389d902f refactor: Drop unused fclose() from BufferedFile
 -:  ---------- > 13:  9999b89cd3 Make BufferedFile to be a CAutoFile wrapper
 -:  ---------- > 14:  fa56c421be Return CAutoFile from BlockManager::Open*File()
 -:  ---------- > 15:  fa4a9c0f43 Remove unused GetType() from OverrideStream, CVectorWriter, SpanReader
 -:  ---------- > 16:  fa72f09d6f Remove CHashWriter type
 -:  ---------- > 17:  fac29a0ab1 Remove SER_GETHASH, hard-code client version in CKeyPool serialize
 -:  ---------- > 18:  96b3f2dbe4 test: add unit test coverage for Python ECDSA implementation
 -:  ---------- > 19:  12f7257b8f doc: Be vague instead of wrong about MALLOC_ARENA_MAX
 -:  ---------- > 20:  b3db8c9d5c rpc: bumpfee, improve doc for 'reduce_output' arg
 -:  ---------- > 21:  4660fc82a1 wallet: Check last block and conflict height are valid in MarkConflicted
 -:  ---------- > 22:  782701ce7d test: Test loading wallets with conflicts without a chain
 -:  ---------- > 23:  b0f5175c04 net: Drop v2 garbage authentication packet
 -:  ---------- > 24:  e3720bca39 net: Simplify v2 recv logic by decoupling AAD from state machine
 -:  ---------- > 25:  fa40b3ee22 test: Avoid test failure on Linux root without cap-add LINUX_IMMUTABLE
 -:  ---------- > 26:  b5a962564e tests: Use manual bumps instead of bumpfee for resendwallettransactions
 -:  ---------- > 27:  d9841a7ac6 Add make_secure_unique helper
 -:  ---------- > 28:  6ef405ddb1 key: don't allocate secure mem for null (invalid) key
 -:  ---------- > 29:  262ab8ef78 Add package evaluation fuzzer
 -:  ---------- > 30:  f9047771d6 lint: fix custom mypy cache dir setting
 -:  ---------- > 31:  380130d9d7 test: add coverage to feature_addrman.py
 -:  ---------- > 32:  d9b172cd00 doc: fix link to developer-notes.md file in multiprocess.md
 1:  7303dac506 = 33:  da384a286b rpc: getrawaddrman for addrman entries
 2:  2dc1a6d59e ! 34:  352d5eb2a9 test: getrawaddrman RPC
    @@ Commit message
         deterministic (yet).

      ## test/functional/rpc_net.py ##
    -@@ test/functional/rpc_net.py: from test_framework.util import (
    - )
    - from test_framework.wallet import MiniWallet
    +@@ test/functional/rpc_net.py: from itertools import product
    + import time

    -+# Address manager size constants as defined in addrman_impl.h
    -+ADDRMAN_NEW_BUCKET_COUNT = 1 << 10
    -+ADDRMAN_TRIED_BUCKET_COUNT = 1 << 8
    -+ADDRMAN_BUCKET_SIZE = 1 << 6
    -+
    -
    - def assert_net_servicesnames(servicesflag, servicenames):
    -     """Utility that checks if all flags are correctly decoded in
    + import test_framework.messages
    ++from test_framework.netutil import ADDRMAN_NEW_BUCKET_COUNT, ADDRMAN_TRIED_BUCKET_COUNT, ADDRMAN_BUCKET_SIZE
    + from test_framework.p2p import (
    +     P2PInterface,
    +     P2P_SERVICES,
     @@ test/functional/rpc_net.py: class NetTest(BitcoinTestFramework):
              self.test_addpeeraddress()
              self.test_sendmsgtopeer()
```

</details>

stratospher commented:
reACK 352d5eb.

achow101 commented:
ACK 352d5eb2a9e89cff4a2815d94a9d81fcc20c4b2c

0xB10C commented on test/functional/rpc_net.py:
```diff
@@ -386,5 +388,115 @@ def test_getaddrmaninfo(self):
+            assert_equal(result["source"], expected["source"])
+            assert_equal(result["source_network"], expected["source_network"])
+            # To avoid failing on slow test runners, use a 10s vspan here.
+            assert_approx(result["time"], time.time(), vspan=10)
```
What's a reasonable time range for slow test runners?

0xB10C replied:
ref https://github.com/bitcoin/bitcoin/pull/28671

0xB10C replied:
ref https://github.com/bitcoin/bitcoin/pull/28671
